tests/framework/crossplane.go

```diff
@@ -39,40 +41,74 @@ type ExpectedNginxField struct {
 // ValidateNginxFieldExists accepts the nginx config and the configuration for the expected field,
 // and returns whether or not that field exists where it should.
 func ValidateNginxFieldExists(conf *Payload, expFieldCfg ExpectedNginxField) error {
+	var directiveFoundInServer, directiveFoundInUpstream bool
+
+	b, err := json.Marshal(conf)
+	if err != nil {
+		return fmt.Errorf("error marshaling nginx config: %w", err)
+	}
+
 	for _, config := range conf.Config {
 		if !strings.Contains(config.File, expFieldCfg.File) {
 			continue
 		}
 
 		for _, directive := range config.Parsed {
-			if len(expFieldCfg.Servers) == 0 {
+			if len(expFieldCfg.Server) == 0 && len(expFieldCfg.Upstream) == 0 {
 				if expFieldCfg.fieldFound(directive) {
 					return nil
 				}
 				continue
 			}
 
-			for _, serverName := range expFieldCfg.Servers {
-				if directive.Directive == "server" && getServerName(directive.Block) == serverName {
-					for _, serverDirective := range directive.Block {
-						if expFieldCfg.Location == "" && expFieldCfg.fieldFound(serverDirective) {
-							return nil
-						} else if serverDirective.Directive == "location" &&
-							fieldExistsInLocation(serverDirective, expFieldCfg) {
-							return nil
-						}
-					}
-				}
+			directiveFoundInServer = validateServerBlockDirectives(expFieldCfg, *directive)
+
+			directiveFoundInUpstream = validateUpstreamDirectives(expFieldCfg, *directive)
+
+			if len(expFieldCfg.Server) > 0 && directiveFoundInServer {
+				return nil
+			}
+
+			if len(expFieldCfg.Upstream) > 0 && directiveFoundInUpstream {
+				return nil
 			}
 		}
 	}
 
-	b, err := json.Marshal(conf)
-	if err != nil {
-		return fmt.Errorf("error marshaling nginx config: %w", err)
+	return fmt.Errorf("directive %s not found in: nginx config %s", expFieldCfg.Directive, string(b))
+}
+
+func validateServerBlockDirectives(
```

Comment:
```suggestion
func fieldExistsInServer(
```

More consistent with the existing location function.